### ISSUE TITLE
[Question][PR-inprogress] Making SelfControl impervious to date/time changes

### ISSUE THREAD

**Issue by shaandsingh** [SelfControlApp/selfcontrol#618]
Hi there, thanks for building this great piece of software. I'm working on fixing #28 by way of #475, and I have a few questions:

- scheckup is being a pain with Swift–ObjC bridging (TrueTime is written in Swift). I know you got rid of it a few days ago. Is it safe to use that commit for this PR? In other words, if I try to run SelfControl with the latest code on master, will it work stably?

- TrueTime needs to be started as close to app start as possible so that there's enough time to make the network request. I chose `applicationDidFinishLaunching`in AppController.m as the spot for this, but are there any other targets which start in other places?

And lastly, a note: I know you've been wary in the past about network calls because of (1) the need for a backup incase the network fails (which TrueTime has some protection against, but not much) and (2) the possibility of a user blocking the NTP server itself (which TrueTime protects against by using a *pool* of servers, though that too isn't a perfect solution because a person could block all of them). My code probably won't solve all of these problems, but I hope that we can get something workable going for the unfortunate set of users who figured out the date/time trick—even if it's just a secondary branch or advanced preferences option.

**Comment by cstigler** [comment]
Hi @shaandsingh, first thanks so much for contributing! It sounds like you've gotten a pretty good hang of the codebase already, which is great. I'd love to help as much as possible with that PR, so feel free to ask any other questions about the codebase or reach out for suggestions. I have a couple specific notes and a couple more general architectural thoughts for you.

Specific responses to your questions:
1. Yes, `scheckup` is gone. The current PR should be safe to build off of, and it also should be stable / releaseable.
2. `applicationDidFinishLaunching` in AppController is a good hook for when the GUI app launches. The helper tools (`org.eyebeam.SelfControl` and `SCKillerHelper`) are both simple programs that run off a single `main` function currently, so you might just insert a call near the top of each file (that'd be `HelperMain.m` and `SCKillerHelper/main.m`, respectively).

Broader thoughts on the NTP protection and your PR:
1. One slight complication is a current change I'm working on. Basically, the current helper tool approach (copying over a privileged helper tool with `AuthorizationExecuteWithPrivileges` and then running it periodically) has been deprecated for about 7 years and is also kinda flaky. I'm moving to the newer recommended approach, which is `SMJobBless` to install a daemon that loops and runs continuously (rather than frequently starting new processes ). I haven't done much on this yet but my work will all be on the `daemon` branch. I'm not sure how much you're planning to touch the helper tools, but you should be aware of that because there will be a lot of code churn coming in the helper tools. Ideally most of the logic is living in separate files anyway (those files are way too big already!) so there's just a few hooks going into the helper tools themselves.

2. One principle that's evolved over the years is: **we should never fail in a way that leaves the user blocked.** It's just not worth the support headache.

3. Instead, the flow I'd prefer is:
  a) if system date/time changes, attempt to seamlessly transfer the block over by adjusting the end date using the TrueTime version
  b) if getting the networked date/time fails for any reason (including the user blocking the NTP server), we trigger the "CHEATER" background by setting `TamperingDetected` to `YES`, but then remove the block / let them get away with it. They still have to deal with the guilt :)

**Comment by shaandsingh** [comment]
Thanks for your comment Charlie; it was very helpful! A few followups:

- At the moment, I think TrueTime will only need to hook in at two places (for the most part): app starts and SCUtilities. Therefore I don't anticipate much conflict with the helper tool upgrades. If that changes though, I'll let you know.

- One problem I forgot to mention in my first message is that TrueTime has a minimum deployment target of 10.10, meaning that it probably makes sense to approach this as an alpha update until a broader solution can be worked out.

- My concern with the cheater background idea is that the user would see it every time the network drops out for a significant period of time, which, for most people and in most occurrences, is innocuous. Imagine working on the train and your hotspot drops out, you're already frustrated because of how cooped up you are, and then a big cheater background takes over your desktop. If we abide by the principle of never leaving the user unfairly blocked, then I imagine we should also abide by its frontend corollary: we should never make a UI change that could unfairly surprise the user.

- Finally, your implementation idea in (3a) is perfect. I'll work on that.

Shaan

**Comment by cstigler** [comment]
Hmm, I thought I replied to this but just looked back and must have forgotten to click the Comment button. To follow up on a few last things:

**Re cheater background:** to be clear, I'm not suggesting we set the background in the case you're describing (working in an offline environment). I think we should only set the background if both of the following conditions are met:

1) there has been an unexpected and significant system timestamp change (_not_ timezone), and
2) the networked time is unavailable.

Case #1 should be very rare - there are few good reasons that system timestamp should change by more than, say, one minute (again, timezone is a whole different deal and shouldn't impact the timestamp). And case #1 happening at the same time as the user being offline - seems unlikely to me. Does that make sense to you?

**Re minimum deployment target:** I'm fine just making the next version target 10.10 and later. We can't be backwards-compatible forever :)

**Comment by shaandsingh** [comment]
Your clarification makes total sense, thanks for that. And we might not have to adjust the minimum deployment target after all—I've made some intriguing discoveries in the last month. Sorry about the delay, I'm working on this feature in the way one strolls along a beautiful path, rather than in a business-critical sense. More soon!

**Comment by cstigler** [comment]
Makes sense! That's how I take all of my SelfControl work, which is why we release so rarely ;)

**Comment by hazelf42** [comment]
Just commenting here and saying I'm really looking forward to this fix since I have no self control of my own 🤦

**Comment by gob35** [comment]
I'm curious- Are you still working on this, @shaandsingh? I know a fix for this issue will make so many people's lives easier (myself included). If you have abandoned this project, would a bounty or even dontations/tips help make it a priority again?

**Comment by slambert** [comment]
Duplicate: https://github.com/SelfControlApp/selfcontrol/issues/28
